From 7a76c7cb6035ec6a4322b87c8aadbdf99fe2fb5b Mon Sep 17 00:00:00 2001
From: Christophe Guibout <christophe.guibout@st.com>
Date: Tue, 24 Oct 2023 16:06:59 +0200
Subject: [PATCH 1/2] plat/st: add management of boot partition

Give next stage boot the boot mode used (0 for normal boot,
2 for recovery).

Change-Id: I0cc406bfcb1f5a631733cf2eba6a3d9fe3400795

# Conflicts:
#	plat/st/stm32mp1/include/stm32mp1_context.h
#	plat/st/stm32mp1/stm32mp1_private.c
---
 plat/st/common/bl2_io_storage.c         |  1 +
 plat/st/common/include/stm32mp_common.h |  6 ++++++
 plat/st/stm32mp1/stm32mp1_private.c     | 17 +++++++++++++++++
 3 files changed, 24 insertions(+)

diff --git a/plat/st/common/bl2_io_storage.c b/plat/st/common/bl2_io_storage.c
index a1e1665c48..67b1bf5b5c 100644
--- a/plat/st/common/bl2_io_storage.c
+++ b/plat/st/common/bl2_io_storage.c
@@ -718,6 +718,7 @@ void plat_fwu_set_images_source(const struct fwu_metadata *metadata)
 	io_block_spec_t *image_spec;
 
 	boot_idx = plat_fwu_get_boot_idx();
+	stm32_save_boot_mode((boot_idx == 1U) ? BOOT_RECOVERY : BOOT_NORMAL);
 	assert(boot_idx < NR_OF_FW_BANKS);
 
 	for (i = 0U; i < NR_OF_IMAGES_IN_FW_BANK; i++) {
diff --git a/plat/st/common/include/stm32mp_common.h b/plat/st/common/include/stm32mp_common.h
index bf015dc02f..c55c4c976d 100644
--- a/plat/st/common/include/stm32mp_common.h
+++ b/plat/st/common/include/stm32mp_common.h
@@ -14,6 +14,11 @@
 #define JEDEC_ST_BKID U(0x0)
 #define JEDEC_ST_MFID U(0x20)
 
+enum boot_mode {
+	BOOT_NORMAL = 0x00,
+	BOOT_RECOVERY = 0x02,
+};
+
 /* Functions to save and get boot context address given by ROM code */
 void stm32mp_save_boot_ctx_address(uintptr_t address);
 uintptr_t stm32mp_get_boot_ctx_address(void);
@@ -136,6 +141,7 @@ uintptr_t stm32_get_bkpr_boot_mode_addr(void);
 void stm32_display_board_info(uint32_t board_id);
 
 #if PSA_FWU_SUPPORT
+int stm32_save_boot_mode(enum boot_mode mode);
 void stm32mp1_fwu_set_boot_idx(void);
 uint32_t stm32_get_and_dec_fwu_trial_boot_cnt(void);
 void stm32_set_max_fwu_trial_boot_cnt(void);
diff --git a/plat/st/stm32mp1/stm32mp1_private.c b/plat/st/stm32mp1/stm32mp1_private.c
index db40a156ba..f4b8d77664 100644
--- a/plat/st/stm32mp1/stm32mp1_private.c
+++ b/plat/st/stm32mp1/stm32mp1_private.c
@@ -33,6 +33,8 @@
 #define TAMP_BOOT_FWU_INFO_IDX_OFF	U(0)
 #define TAMP_BOOT_FWU_INFO_CNT_MSK	GENMASK(7, 4)
 #define TAMP_BOOT_FWU_INFO_CNT_OFF	U(4)
+#define TAMP_BOOT_MODE_MASK		GENMASK(7, 0)
+#define TAMP_BOOT_MODE_SHIFT		0
 
 #if defined(IMAGE_BL2)
 #define MAP_SEC_SYSRAM	MAP_REGION_FLAT(STM32MP_SYSRAM_BASE, \
@@ -722,6 +724,21 @@ uintptr_t stm32_get_bkpr_boot_mode_addr(void)
 }
 
 #if PSA_FWU_SUPPORT
+int stm32_save_boot_mode(enum boot_mode mode)
+{
+	uint32_t bkpr_itf_idx = tamp_bkpr(TAMP_BOOT_MODE_BACKUP_REG_ID);
+
+	clk_enable(RTCAPB);
+
+	mmio_clrsetbits_32(bkpr_itf_idx,
+			   TAMP_BOOT_MODE_MASK,
+			   mode << TAMP_BOOT_MODE_SHIFT);
+
+	clk_disable(RTCAPB);
+
+	return 0;
+}
+
 void stm32mp1_fwu_set_boot_idx(void)
 {
 	clk_enable(RTCAPB);
-- 
2.25.1

